olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.h

```diff
@@ -19,25 +19,50 @@
 
 #pragma once
 
+#include <memory>
+
+#include <olp/core/client/CancellationContext.h>
+#include <olp/core/client/CancellationToken.h>
 #include <olp/core/client/HRN.h>
 #include <olp/core/client/OlpClientSettings.h>
-#include <olp/dataservice/read/Types.h>
+#include <olp/core/client/PendingRequests.h>
+#include <olp/dataservice/read/SubscribeRequest.h>
+#include "repositories/ExecuteOrSchedule.inl"
 
 namespace olp {
 namespace dataservice {
 namespace read {
 
-class StreamLayerClientImpl {
+class StreamLayerClientImpl
+    : public std::enable_shared_from_this<StreamLayerClientImpl> {
  public:
   StreamLayerClientImpl(client::HRN catalog, std::string layer_id,
                         client::OlpClientSettings settings);
 
   virtual ~StreamLayerClientImpl();
 
+  virtual bool CancelPendingRequests();
+
+  virtual client::CancellationToken Subscribe(
+      SubscribeRequest request,
+      SubscribeRequest::SubscribeResponseCallback callback);
+
+  virtual client::CancellableFuture<SubscribeRequest::SubscribeResponse>
+  Subscribe(SubscribeRequest request);
+
  private:
   client::HRN catalog_;
   std::string layer_id_;
   client::OlpClientSettings settings_;
+  std::shared_ptr<client::PendingRequests> pending_requests_;
+
+  std::mutex mutex_;
+
+  boost::optional<std::string> subscription_id_;
+  boost::optional<std::string> subscription_mode_;
+  boost::optional<std::string> consumer_id_;
+  boost::optional<std::string> x_correlation_id;
+  boost::optional<std::string> node_base_url_;
```

Comment:
Better to replace those fields with single struct.

Reply:
Not sure why we need that. What do we achieve?

Reply:
That was also my proposal in the design document.
We achieve better design and act on the single responsibility principle.